Quantization and Binarization Implementations for Linear Layers

mmoffatt2 opened the pull request:
Following code from [Quantization Repo](https://github.com/Alexstrasza98/Transformer-Quantization/blob/main/quantization/quantize.py), integrated quantization into the nanoGPT repo. 

Added a new QuantizedLinear layer that replaced linear layers and simulates quantization aware training. Also, added a binarization layer.

gkielian commented:
Awesome! This trains and converges!

We should try what the output from quantized inference looks like with the sample.py.
When trying to run sample.py on the created checkpoints, it encounters unexpected keys in the state_dict:
![image](https://github.com/ReaLLMASIC/nanoGPT/assets/10952473/d2f8179a-7153-470b-a422-e972aa748714)

Could you take a look into how to test with the sample.py (feel free to modify if needed)?

gkielian commented on gpt_conf.py:
```diff
@@ -11,6 +11,9 @@ class GPTConfig:
     gate: bool = False
+    quantization_choice: str = "none" # Choices: "none" "quantize" "binarize" "irbinarize"
+    quantization_bits: int = 8
+    quantize_attention: bool = True
```
In order to increase the combinations we can try, let's separate the different types of and levels of quantization.

Seeing the results from the llama.cpp and GGUF format communities, we'll probably be able to succeed with more aggressive quantization from the "Wk" and "Wq" matrices than "Wv" matrices. 

Keep one global quantization, maybe "quantize_attn_all", which if not None sets everything to one quantization, and otherwise would default to granular quantization settings:

quantize_attn_all = None
quantize_attn_q = None
quantize_attn_k = None
quantize_attn_v = None
quantize_attn_proj = None

Similarly for MLP:
quantize_mlp_all = None
quantize_mlp_up = None
quantize_mlp_down = None

gkielian commented on gpt_conf.py:
```diff
@@ -11,6 +11,9 @@ class GPTConfig:
     dropout: float = 0.0
     window_size: int = 128
     gate: bool = False
+    quantization_choice: str = "none" # Choices: "none" "quantize" "binarize" "irbinarize"
```
Let's utilize the built-in None instead of string "none"

gkielian commented on model.py:
```diff
@@ -82,12 +85,16 @@ def create_shared_param_group(layer_type, config):
         assert config.n_embd % config.n_head == 0
         # key, query, value projections for all heads, but in a batch
-        self.c_attn_q = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
+        if config.quantization_choice == 'quantize' and config.quantize_attention:
```
I see, adding the modifications here is a solid approach.

To prevent a long if else, utilize the linear_dictionary with the mappings (see end of variations/linear_variations)

We'll want to create polymorphism between the different alternatives, so use
Quantized_Linear( config.n_embd, config.n_embd, config=config) as the function signatures, using the config parameter in the  __init__ of each to set the self.bias and self.quantization_bits for each class (if needed).

Handle 
1. Polymorphism amongst the

gkielian commented on model.py:
```diff
@@ -97,10 +104,20 @@ def __init__(self, config, fire_pos_enc=None):
-        self.c_attn_v = nn.Linear(config.n_embd, self.kv_dim, bias=config.bias)
-        # output projection
-        self.c_proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
+        if config.quantization_choice == 'quantize' and config.quantize_attention:
+            self.c_attn_k = QuantizedLinear(config.quantization_bits, config.n_embd, self.kv_dim, bias=config.bias)
+            self.c_attn_v = QuantizedLinear(config.quantization_bits, config.n_embd, self.kv_dim, bias=config.bias)
+            self.c_proj = QuantizedLinear(config.quantization_bits, config.n_embd, config.n_embd, bias=config.bias)
+        elif config.quantization_choice == 'binarize':
+            self.c_attn_k = BinarizedLinear(config.n_embd, self.kv_dim, bias=True)
+            self.c_attn_v = BinarizedLinear(config.n_embd, self.kv_dim, bias=True)
+            self.c_proj = BinarizedLinear(config.n_embd, config.n_embd, bias=True)
+        else:
+            self.c_attn_k = nn.Linear(config.n_embd, self.kv_dim, bias=config.bias)
+            self.c_attn_v = nn.Linear(config.n_embd, self.kv_dim, bias=config.bias)
+            # output projection
+            self.c_proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
+
```
Great, and reduce the size using dictionary and polymorphism

gkielian commented on model.py:
```diff
@@ -257,20 +274,40 @@ def __init__(self, config):
 
         # Select linear variant
         self.linear_variant = linear_dictionary[config.linear_variant]
+        if config.quantization_choice == 'quantize':
+            self.linear_variant = QuantizedLinear
```
After adding to the linear_dictionary can save these four lines.

gkielian commented on model.py:
```diff
@@ -257,20 +274,40 @@ def __init__(self, config):
-            self.c_fc_out = linear_dictionary[config.linear_variant](4 * config.n_embd, config.n_embd, bias=config.bias)
+            if config.quantization_choice == 'quantize':
+                self.c_fc_in1 = self.linear_variant(config.quantization_bits, config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_fc_in2 = self.linear_variant(config.quantization_bits, config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_fc_out = self.linear_variant(config.quantization_bits, 4 * config.n_embd, config.n_embd, bias=config.bias)
+            elif config.quantization_choice == 'binarize':
+                self.c_fc_in1 = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=True)
+                self.c_fc_in2 = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=True)
+                self.c_fc_out = self.linear_variant(4 * config.n_embd, config.n_embd, bias=True)
+            else:
+                self.c_fc_in1 = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_fc_in2 = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_fc_out = self.linear_variant(4 * config.n_embd, config.n_embd, bias=config.bias)
         else:
-            self.c_fc = linear_dictionary[config.linear_variant](config.n_embd, 4 * config.n_embd, bias=config.bias)
-            self.c_proj = linear_dictionary[config.linear_variant](4 * config.n_embd, config.n_embd, bias=config.bias)
+            if config.quantization_choice == 'quantize':
+                self.c_fc = self.linear_variant(config.quantization_bits, config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_proj = self.linear_variant(config.quantization_bits, 4 * config.n_embd, config.n_embd, bias=config.bias)
+            elif config.quantization_choice == 'binarize':
+                self.c_fc = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=True)
+                self.c_proj = self.linear_variant(4 * config.n_embd, config.n_embd, bias=True)
+            else:
+                self.c_fc = self.linear_variant(config.n_embd, 4 * config.n_embd, bias=config.bias)
+                self.c_proj = self.linear_variant(4 * config.n_embd, config.n_embd, bias=config.bias)
```
Use  polymorphism and linear_dictionary for this section also

gkielian commented on quantization/quantize.py:
Keep the common quantization features in this file, and move the Linear variations to the variations/linear_variations folder.

Then add this file as an import to the linear_variations.py,so that we can reuse this code as a basis for future quantization of activation functions, normalization operations, and softmax alternatives

gkielian commented on sample.py:
Excellent! what would be a good sample command for running this be? Would you be able to add a section in the README.md with a few lines for demonstration (python3 data/shakespeare_char/prepare.py; python3  train.py; python3 sample.py --argparse args)  with an image sample output?

gkielian reviewed (changes requested):
This is great, and looks like it can serve as the foundation for future quantization with  the other elements of the networks.

Please see comments in review and definitely reach out if there are questions on implementation of the remaining modifications.

mmoffatt2 commented:
Unfortunately, I didn't think ahead to split these changes into multiple PRs. I will try to summarize all the changes as best as possible.

Updates:
Added two options for quantization: "affine_quant" and "stochastic_quant"
Added QuantizedEmbedding to position_encoding_variations.py, which can be used for word embeddings and position embeddings
Added quantize_helper function which is used for quantization of the outputs of matrix multiplications and linear layers

train.py arguments:
"quantization_linear_method": quantization method used for all quantized linear layers
"quantize_wte": whether the word embedding weights are quantized (using QuantizedEmbedding)
"quantize_wpe": whether the position embedding weights are quantized (using QuantizedEmbedding)
"quantization_embedding_method": method of quantization used for embeddings
"quantize_attn_all": this boolean argument will quantize everything in the attn layer, including outputs and linear layers
"quantize_c_attn_q": whether the query weights are quantized (using QuantizedLinear)
"quantize_c_attn_k": whether the key weights are quantized (using QuantizedLinear)
"quantize_c_attn_v": whether the value weights are quantized (using QuantizedLinear)
"quantize_q": whether the query output is quantized (using quantize_helper)
"quantize_k": whether the key output is quantized (using quantize_helper)
"quantize_v": whether the value output is quantized (using quantize_helper)
"quantize_q_k_mult": whether the output of the query key matrix multiplication is quantized (using quantize_helper)
"quantize_softmax_v_mult": whether the output of the softmax value matrix multiplication is quantized (using quantize_helper)
"quantize_softmax": whether the output of the softmax variant is quantized (using quantize_helper)
"quantize_attn_proj": whether the output projection is quantized (using QuantizedLinear)
"quantize_mlp_all": this boolean argument will quantize everything in the mlp layer, including activation outputs and linear layers
"quantize_mlp_up": whether the mlp_up weights are quantized (using QuantizedLinear)
"quantize_mlp_down": whether the mlp_down weights are quantized (using QuantizedLinear)
"quantize_activation": whether the output of the activation function is quantized (using quantize_helper)
"quantization_activation_method": method of quantization for the activation output

Example code for running train.py:
python3 train.py --device="cuda" --dataset="shakespeare_char" --out_dir="out_shakespeare_char" --linear_variant="quantized_linear" --quantize_attn_all --quantize_mlp_all --quantize_wte --quantize_wpe

Example code for running sample.py:
python3 sample.py --out_dir="out_shakespeare_char" --device="cuda" --quant_weights_file="shakespeare" --visualize_weights_dir="shakespeare_weights"

gkielian commented:
Will start diving into the code. one thing we'll have to separate out the quantization for the attention output projection -- experiment with this show that it might need it's own unique precision (possibly higher) in order to operate well.
